consensus/src/consensus_observer/network/observer_message.rs

```diff
@@ -674,33 +675,64 @@ impl BlockPayload {
-                |error| {
-                    Error::InvalidMessageError(format!(
-                        "Failed to verify payload proof digests! Num transactions: {:?}, \
+            match reconstruct_batch(
+                &block_info,
+                &mut transactions_iter,
+                proof_of_store.info(),
+                true,
+            ) {
+                Ok(Some(batch_transactions)) => {
+                    batches_and_transactions
+                        .push((proof_of_store.info().clone(), batch_transactions));
+                },
+                Ok(None) => { /* Nothing needs to be done (the batch was expired) */ },
+                Err(error) => {
+                    return Err(Error::InvalidMessageError(format!(
+                        "Failed to reconstruct payload proof batch! Num transactions: {:?}, \
                         num batches: {:?}, num inline batches: {:?}, failed batch: {:?}, Error: {:?}",
                         num_transactions, num_payload_proofs, num_inline_batches, proof_of_store.info(), error
-                    ))
+                    )));
                 },
-            )?;
+            }
         }
 
-        // Verify the inline batch digests using the transactions
+        // Collect the transactions for each inline batch
         for batch_info in inline_batches.into_iter() {
-            reconstruct_and_verify_batch(&block_info, &mut transactions_iter, batch_info, false).map_err(
-                |error| {
-                    Error::InvalidMessageError(format!(
-                        "Failed to verify inline batch digests! Num transactions: {:?}, \
+            match reconstruct_batch(&block_info, &mut transactions_iter, batch_info, false) {
+                Ok(Some(batch_transactions)) => {
+                    batches_and_transactions.push((batch_info.clone(), batch_transactions));
+                },
+                Ok(None) => {
+                    return Err(Error::UnexpectedError(format!(
+                        "Failed to reconstruct inline batch! Batch was unexpectedly skipped: {:?}",
+                        batch_info
+                    )));
+                },
+                Err(error) => {
+                    return Err(Error::InvalidMessageError(format!(
+                        "Failed to reconstruct inline batch! Num transactions: {:?}, \
                         num batches: {:?}, num inline batches: {:?}, failed batch: {:?}, Error: {:?}",
                         num_transactions, num_payload_proofs, num_inline_batches, batch_info, error
-                    ))
+                    )));
                 },
-            )?;
+            }
         }
 
+        // Verify all the reconstructed batches (in parallel)
+        batches_and_transactions
+            .into_par_iter()
+            .try_for_each(|(batch_info, transactions)| verify_batch(&batch_info, transactions))
```

Comment:
nit: we may want to have a min_len to avoid too many small tasks

Reply:
Done boss 😄